ignite/engine/engine.py

```diff
@@ -371,7 +372,12 @@ def _fire_event(self, event_name: Any, *event_args, **event_kwargs) -> None:
             self.last_event_name = event_name
             for func, args, kwargs in self._event_handlers[event_name]:
                 kwargs.update(event_kwargs)
-                func(self, *(event_args + args), **kwargs)
+                signature = inspect.signature(func)
+                try:
+                    signature.bind(self, *(event_args + args), **kwargs)
+                    func(self, *(event_args + args), **kwargs)
+                except TypeError:
+                    func(*(event_args + args), **kwargs)
```

Comment:
@sdesrozis can't we inspect and setup *args with or without engine at `add_event_handler` time and here just trigger the handler without dispatch ?

Reply:
See the other comment.

I would have preferred that. But exception arguments could be added by `_fire_event`. If we agree the order of argument (`engine` not always the first), we could do what you said (and I prefer that!)

Reply:
We can keep the order as it is currently: Engine should be the first argument.

Reply:
Consider this pseudo code
```
engine.add_event_handler(event, fn, *args, **kwargs)
```
2 alternatives :
- `fn` doesn’t have engine in signature, we keep `(fn,*args, **kwargs)` for `event`
- `fn` has engine in signature, we keep `(fn,(engine, *args), **kwargs)` for `event`
From now, we shouldn’t have no longer dealing with `engine`. 
But consider 
```
_fire_event(event, *event_args, **event_kwargs)
```
We have to merge arguments we have kept for `event` and the arguments mais passed to the function `_fire_event`.

Reply:
@sdesrozis I think we still can store `(fn, *args, **kwargs)` at `add_event_handler`:
```python
class Engine:
    
    def __init__(self):
        self.hs = []


def add(engine, fn, *args, **kwargs):
    
    signature = inspect.signature(fn)
    
    event_args = ()
    event_kwargs = {}
        
    try:
        signature.bind(engine, *args, **kwargs)
        args = (engine, ) + event_args + args
        kwargs.update(event_kwargs)
    except TypeError as e:
        try:
            signature.bind(*args, **kwargs)
            args = event_args + args
            kwargs.update(event_kwargs)
        except TypeError as e:
            raise RuntimeError("Wrong signature...")            

    engine.hs.append((fn, args, kwargs))


def trigger(engine):
    for h in engine.hs:
        fn = h[0]
        fn(*h[1], **h[2])
```
What do you think ?

Anyway, we will need to provide extensive tests to cover possible cases.

Reply:
I exactly did that :)